src/Microsoft.AspNet.Server.Kestrel/Http/PathNormalizer.cs

```diff
@@ -0,0 +1,132 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+
+namespace Microsoft.AspNet.Server.Kestrel.Http
+{
+    public static class PathNormalizer
+    {
+        public static string Normalize(string path)
+        {
+            if (ContainsDotSegments(path))
+            {
+                var normalizedChars = new char[path.Length];
```

Comment:
```csharp
using System.Buffers;
...
var normalizedChars = ArrayPool<char>.Shared.Rent(path.Length);
// stuff
ArrayPool<char>.Shared.Return(normalizedChars);
```

project.json
```json
"System.Buffers": "4.0.0-*"
```

Doesn't need try catch for the return; will just be GC'd if you don't return it.